### PR TITLE
Story(CCLS-2083&CCLS-2084) Refactoring error handlers and lookup service.

### PR DESCRIPTION
**Pull request by PhilDigitalJustice**

### CODE REVIEW

File: src/main/java/uk/gov/laa/ccms/caab/client/CaabApiClient.java [hunk 1]
```diff
@@ -33,7 +33,58 @@ public class CaabApiClient {
 
   private final WebClient caabApiWebClient;
 
-  private final CaabApiClientErrorHandler caabApiClientErrorHandler;
+  private final ApiClientErrorHandler apiClientErrorHandler;
+
+  public static final String FAILED_CREATE_APPLICATION =
```
Comment (porritta) [inline src/main/java/uk/gov/laa/ccms/caab/client/CaabApiClient.java:38]: I think maybe you were right to suggest refactoring further to remove the need for all these constants. As per my slack msg.

File: src/main/java/uk/gov/laa/ccms/caab/bean/validators/costs/CostDetailsValidator.java [hunk 2]
```diff
@@ -4,6 +4,7 @@
 import org.springframework.stereotype.Component;
 import org.springframework.validation.Errors;
 import uk.gov.laa.ccms.caab.bean.costs.CostsFormData;
+import uk.gov.laa.ccms.caab.bean.priorauthority.PriorAuthorityFormDataDetails;
```
Comment (porritta) [inline src/main/java/uk/gov/laa/ccms/caab/bean/validators/costs/CostDetailsValidator.java:7]: This is the only change in this file, so presume it's unused.

Comment (PhilDigitalJustice) [reply src/main/java/uk/gov/laa/ccms/caab/bean/validators/costs/CostDetailsValidator.java:7]: amended

File: src/main/java/uk/gov/laa/ccms/caab/client/CaabApiClient.java (hunk 1 above)
Comment (PhilDigitalJustice) [reply src/main/java/uk/gov/laa/ccms/caab/client/CaabApiClient.java]: amended

File: src/main/java/uk/gov/laa/ccms/caab/client/AbstractApiClientErrorHandler.java [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+public abstract class AbstractApiClientErrorHandler {
+
+  /**
+   * Creates an exception with a specified message and cause.
+   *
+   * @param message the detail message for the exception.
+   * @param cause the cause of the exception.
+   * @return a new instance of RuntimeException.
+   */
+  protected RuntimeException createException(final String message, final Throwable cause) {
```
Comment (porritta) [inline src/main/java/uk/gov/laa/ccms/caab/client/AbstractApiClientErrorHandler.java:22]: Do you not want to make this method abstract, so it has to be defined in the specific error handlers?

Comment (PhilDigitalJustice) [reply src/main/java/uk/gov/laa/ccms/caab/client/AbstractApiClientErrorHandler.java:22]: makes sense, will amend

File: src/main/java/uk/gov/laa/ccms/caab/client/CaabApiClient.java [hunk 4]
```diff
@@ -576,8 +612,8 @@ public Mono<Void> deletePriorAuthority(
         .header("Caab-User-Login-Id", loginId)
         .retrieve()
         .bodyToMono(Void.class)
-        .onErrorResume(e -> caabApiClientErrorHandler
-            .handleUpdatePriorAuthorityError(e, priorAuthorityId));
+        .onErrorResume(e -> caabApiClientErrorHandler.handleApiUpdateError(e,
+            RESOURCE_TYPE_PRIOR_AUTHORITIES, "id", priorAuthorityId.toString()));
```
Comment (porritta) [inline src/main/java/uk/gov/laa/ccms/caab/client/CaabApiClient.java:616]: String.valueOf(priorAuthorityId)

**Review by porritta — CHANGES_REQUESTED** [review]
Comments added

File: src/main/java/uk/gov/laa/ccms/caab/client/CaabApiClient.java [hunk 5]
```diff
@@ -555,8 +591,8 @@ public Mono<Void> updatePriorAuthority(
         .bodyValue(data)
         .retrieve()
         .bodyToMono(Void.class)
-        .onErrorResume(e -> caabApiClientErrorHandler
-            .handleUpdateProceedingError(e, priorAuthorityId));
+        .onErrorResume(e -> caabApiClientErrorHandler.handleApiUpdateError(e,
+            RESOURCE_TYPE_PRIOR_AUTHORITIES, "id", priorAuthorityId.toString()));
```
[the remark this answers is not in the document]
Comment (PhilDigitalJustice) [reply src/main/java/uk/gov/laa/ccms/caab/client/CaabApiClient.java:595]: amended

File: src/main/java/uk/gov/laa/ccms/caab/client/CaabApiClient.java (hunk 4 above)
Comment (PhilDigitalJustice) [reply src/main/java/uk/gov/laa/ccms/caab/client/CaabApiClient.java:616]: amended

**Review by porritta — APPROVED** [review]
LGTM

**Pull request merged**